### DISCUSSION TITLE
The inferred type of this node exceeds the maximum length the compiler will serialize.

### DISCUSSION THREAD

**Discussion by HamzaAliArgon** [trpc/trpc#5636] [Q&A]
I'm facing this issue. The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed on the app router type.
I have tried merge routers but it's still not working 
If someone has encountered this issue and resolved this please let me know.
In the future, I have to add more routes in my app but I'm stuck here. 
I also tried some generic types but that is causing the issue in the react app context.
currently, it's working with 88 routes (including all the nested routes t.router)

**Comment by TheHinky** [comment]
This doesn't help. But i am struggling with this big time too, if anyone can provide some sort of help would be appreciated big time :)

**Comment by BeBoRE** [comment]
What have you tried? Could take a look at this:
https://stackoverflow.com/questions/68463963/typescript-the-inferred-type-of-this-node-exceeds-the-maximum-length-the-compi

**Reply by BeBoRE** [reply]
Maybe you could remove the size limit?
https://www.typescriptlang.org/tsconfig#disableSizeLimit

**Reply by benevbright** [reply]
`disableSizeLimit` option doesn't work for me.

**Comment by TheHinky** [comment]
Okay so this is what has fixed it for me. (reason I generating types is to help with auto complete speed, I am hoping. But ill need to check that over next few days)

I saw recommendations on generating types to speed up auto complete. However I was getting this error: 
The inferred type of this node exceeds the maximum length the compiler will serialize

I just had another crack at it to fix it. And for me, the issue was my prisma in context of the trpc. 
For reference, my prisma scheme file is about 2200 lines long, so id say large. 
I didn't notice until now but in the type creation for context, whole prisma schema was getting typed out everywhere. I mean everywhere. 

I've removed prisma from context and just importing prisma directly whenever i need it. (not sure of down sides)

I cant share much code, just basic example. 
I have a business log router, extremely simple: 
![image](https://github.com/trpc/trpc/assets/50818243/420cf134-f396-454b-88d0-3075dfd9e23b)

This is  producing a type file (with prisma still in context), of 3542 lines long for this basic bus log router. 
Prisma type model is shown up 6 times in this one type file, each time its referenced its about 561 lines long. that's 3366 of type just being prisma because its in context. 

After I've removed prisma from context: 
This is the finished type file total lines 102, from 3542. 

(now this issue would of caused hundreds of thousands of extra lines just because of prisma for my project for types)

So that will be the maximum length limit for my project. 
I am not sure of down sides, I'm thinking there wont be any and hopefully, biggest thing is greatly increase auto complete performance. Which ill test over the coming days.

**Reply by andrewdoro** [reply]
hey, can you try implementing this pattern, in my case I had the same problem when using a drizzle client with a lot of inference. Started using `satisfies` for all my routers and now I only get a generic type with the router response type.
https://github.com/t3-oss/create-t3-turbo/blob/main/packages/api/src/router/post.ts#L40

**Reply by TheHinky** [reply]
thank you, ill give that s shot sometime soon, and let you know if it worked for me too.

**Reply by j-fdion** [reply]
> hey, can you try implementing this pattern, in my case I had the same problem when using a drizzle client with a lot of inference. Started using `satisfies` for all my routers and now I only get a generic type with the router response type. https://github.com/t3-oss/create-t3-turbo/blob/main/packages/api/src/router/post.ts#L40

This worked nicely for me, thank you!

**Reply by adambarito** [reply]
Thanks so much for this! I took my drizzle out of the trpc context and everything is so much faster!

**Comment by statusunknown418** [comment]
something that also works is typing your function return

**Comment by sudo-vaibhav** [comment]
for anyone struggling with this, you might want to also look into potential bloat in your trpc context.

in my case i removed, ...userAccessControlData and went for a more targeted and leaner object (you can also look into complex types being simplified , like replacing literals with normal strings instead of removing keys altogether from context object)

below is how i refactored my context populator function:

```
return {
    // this object is shared across trpc routes, be very careful with trpc context
    // bloat since that can kill typescript performance, as well as cause tsc transpilation errors
    // ...userAccessControlData,
    ability: userAccessControlData.ability,
    rules: userAccessControlData.rules,
    user: {
      hasOneOfRoles: userAccessControlData.user.hasOneOfRoles,
      hasManagerRole: userAccessControlData.user.hasManagerRole,
      partner: userAccessControlData.user.partner
    },
    container,
    securedTenant,
    tenant,
    defaultChannel: channel,
    tenantQueueManager,
    auditLogRepository,
    forbiddenError: ForbiddenError.from(userAccessControlData.ability)
  };
  ```
